simulator/func_sim/operation.h

```diff
@@ -66,21 +68,31 @@ class Operation
 	//target is known at ID stage and always taken
 	bool is_direct_jump() const { return operation == OUT_J_JUMP; }
 
+	//target is known at ID stage; likely to be taken
+	bool is_likely_branch() const { return operation == OUT_BRANCH_LIKELY; }
+
 	//target is known at ID stage but if branch is taken or not is known only at EXE stage
 	bool is_branch() const { return operation == OUT_BRANCH; }
 
 	// target is known only at EXE stage
 	bool is_indirect_jump() const { return operation == OUT_R_JUMP; }
 
-	bool is_jump() const { return this->is_direct_jump()     ||
-				      this->is_branch()   ||
-				      this->is_indirect_jump(); }
+	bool is_jump() const { return this->is_direct_jump()
+	                           || this->is_branch()
+	                           || this->is_likely_branch()
+	                           || this->is_indirect_jump(); }
 
     bool is_taken() const
     {
         return ( this->is_direct_jump() ) || ( this->is_indirect_jump() ) || is_taken_branch;
     }
 
+    bool is_misprediction( const BPInterface& bp_data) const
+    {
+	    return ( is_direct_jump() && ( !bp_data.is_taken || bp_data.target != get_decoded_target()))
+              || is_likely_branch();
+    }
```

Comment:
Placing that function to the class may be OK for that case, but it is bad as a long-term solution.
The issue is that you're bringing _specific_ behavior of decoder stage to _common_ scope of visibility. It leads to inclusion of additional structures, like BP interface etc. It is harmless now, but in the future it may introduce additional coupling.

The usual programming pattern is to make this method external to the class and put it to _specific_ scope (decoder.cc):

```c++
template<typename Instr>
bool is_misprediction( const Instr& instr, const BPInterface& bp_data)
{
    return (instr.is_direct_jump() // ...
}
```

Reply:
In other words, Operation/FuncInstr/PerfInstr is not a class — it is a data structure. The principle difference is that data structure is processed mostly externally